Rewrite net.box state machine in C

locker opened the pull request:
This PR rewrites the net.box state machine in C. This is a prerequisite for adding TLS support to net.box.

Closes #6291
Closes #6617

Gerold103 commented on src/box/lua/net_box.c:
```diff
@@ -185,7 +185,6 @@ static const char netbox_request_typename[] = "net.box.request";
  * We keep a reference to each C function that is frequently called with
```
Why do you need this patch? It does not look right to terminate the connection when one of triggers for one of public events fails. It is not critical for the connection. It is like terminating the connection when `conn.space.test:insert` fails. Please, lets leave it working as before.

locker replied:
It won't terminate the connection if one of the triggers registered for a public event fails, because we don't call the user callbacks directly here. Instead we call a dispatcher function that invokes the callbacks in a fiber, under a pcall, so it can only fail on a memory allocation error so pcall is redundant here.

https://github.com/tarantool/tarantool/blob/e0e6fe106d11855631a2673dbc72a670378669aa/src/box/lua/net_box.lua#L766-L780

https://github.com/tarantool/tarantool/blob/e0e6fe106d11855631a2673dbc72a670378669aa/src/box/lua/net_box.lua#L934-L968

Gerold103 reviewed (changes requested):
Thanks for the patchset!

locker commented on src/box/lua/net_box.lua:
```diff
@@ -174,13 +174,9 @@ local function on_push_sync_default() end
---
-local function create_transport(host, port, user, password, callback)
+local function create_transport(host, port, user, password, callback,
+                                connect_timeout, reconnect_after)
```
Far not all options are needed in this function. I'd pass all required parameters explicitly. If we add a new option affecting the state machine, we'll need to patch the state machine code anyway to support this new option. There's nothing wrong to patch this function signature along the way IMO.

locker commented on src/box/lua/net_box.c:
```diff
@@ -2053,22 +2041,16 @@ netbox_iproto_auth(struct lua_State *L)
 	struct xrow_header hdr;
 	if (netbox_send_and_recv(transport, fd, &hdr) != 0)
-		goto error;
+		return luaT_error(L);
```
It is consistent: functions that deal with a Lua state raise errors while low-level functions that don't use a Lua state return -1.

The reason why I'm doing this is that this change makes it easier to convert everything to C. We have two options here:
1. Either wrap all Lua code pieces in pcalls and return -1 if any of them fails.
2. Wrap the whole state machine code in a pcall.

The latter seems easier and requires less updates of the existing functions. Take `netbox_transport_dispatch_response` for example. It calls some Lua functions and creates some Lua objects - wrapping all those pieces in pcalls would make the code look convoluted IMO, it's easier to wrap the whole function in a pcall. Then, this function is called by `netbox_iproto_schema`, which also pushes some values to a Lua stack, so going further, we can wrap the whole `netbox_iproto_schema` function in a pcall to keep the code linear and simple, and so on.

I'm currently working on the patch converting the state machine to C. I'll update this patch set soon with it so that you can see the whole picture.

locker commented on src/box/lua/net_box.lua:
```diff
@@ -154,14 +154,12 @@ local function on_push_sync_default() end
---                                             -> [error]
---  (any_state, but [error]) -> [closed]
+--                                            -> error
+--  (any state, but 'error') -> closed
```
Done.

Gerold103 commented on src/box/lua/net_box.c:
```diff
@@ -2114,40 +2374,181 @@ netbox_iproto_schema(struct lua_State *L)
+	assert(transport->self_ref = LUA_NOREF);
+	transport->L = luaT_newthread(L);
+	if (transport->L == NULL)
+		return luaT_error(L);
```
If you throw the error anyway, you can use `lua_newthread` as is.

locker replied:
Right. Fixed. Thanks!

locker commented on src/box/lua/net_box.c:
```diff
@@ -2114,40 +2374,181 @@ netbox_iproto_schema(struct lua_State *L)
+ * Connection handler. Raises a Lua error on termination.
+ */
+static int
+netbox_connection_handler_f(struct lua_State *L)
```
Applied. It looks better indeed. Thanks!

locker commented on src/box/lua/net_box.c:
```diff
@@ -100,19 +113,125 @@ enum netbox_method {
+	/** Fiber doing I/O and dispatching responses. */
+	struct fiber *worker;
+	/** Lua thread used by the worker fiber or NULL. */
+	struct lua_State *L;
```
Agree. The transport callback may need a state (e.g. for running on_connect triggers). Better set `storage.lua.stack` so that it can reuse the existing state instead of creating a new one. Done. Thanks!

locker commented on src/box/lua/net_box.c:
```diff
@@ -2038,7 +2038,7 @@ netbox_iproto_id(struct lua_State *L)
 /**
```
Makes sense. Done.

locker commented on src/box/lua/net_box.c:
```diff
@@ -1750,34 +1922,57 @@ netbox_new_transport(struct lua_State *L)
+	if (transport->state != NETBOX_ACTIVE &&
+	    transport->state != NETBOX_FETCH_SCHEMA) {
+		if (transport->last_error != NULL) {
+			diag_set_error(diag_get(), transport->last_error);
```
Right. Let's ignore this for now. I filed a separate ticket to fix this: https://github.com/tarantool/tarantool/issues/6632. Will do as a follow-up.

Gerold103 commented:
Thanks for the fixes! The code looks good now, but something is wrong with how it works in vshard - `misc/check_uuid_on_connect.test.lua` and `router/router.test.lua` fail with your last commit and pass before it. Some other tests also fail, but that was caused by my patch changing the message of `box.error.READ_ONLY`, ignore them.

locker commented:
> Thanks for the fixes! The code looks good now, but something is wrong with how it works in vshard - `misc/check_uuid_on_connect.test.lua` and `router/router.test.lua` fail with your last commit and pass before it. Some other tests also fail, but that was caused by my patch changing the message of `box.error.READ_ONLY`, ignore them.

- `router/router.test.lua` was failing, because vshard expects an `ER_NO_CONNECTION` error from net.box. I updated `luaT_netbox_transport_make_request` to set a new `ClientError` error instead of reusing `transport->last_error`, as it was done before in Lua. BTW it also addresses your concern about possible `last_error` modifications so I removed #6632 from my todo list for now.
- `misc/check_uuid_on_connect.test.lua` was failing, because I forgot to add `luaL_testcancel` to the transport stop method (before the patch, the check was done implicitly by `fiber.cancel` in Lua). As a result, `conn:close` wouldn't raise an error if called on an `on_connect` trigger, which resulted in invalid connection state (connection state is updated *after* running triggers so not failing on close would result in setting the state to 'active').

<details><summary>diff</summary>

```diff
diff --git a/src/box/lua/net_box.c b/src/box/lua/net_box.c
index e3882be5ce3c..34c0b7788d72 100644
--- a/src/box/lua/net_box.c
+++ b/src/box/lua/net_box.c
@@ -1932,8 +1932,9 @@ luaT_netbox_transport_make_request(struct lua_State *L, int idx,
 {
 	if (transport->state != NETBOX_ACTIVE &&
 	    transport->state != NETBOX_FETCH_SCHEMA) {
-		if (transport->last_error != NULL) {
-			diag_set_error(diag_get(), transport->last_error);
+		struct error *e = transport->last_error;
+		if (e != NULL) {
+			box_error_raise(ER_NO_CONNECTION, e->errmsg);
 		} else {
 			const char *state = netbox_state_str[transport->state];
 			box_error_raise(ER_NO_CONNECTION,
@@ -2540,8 +2541,11 @@ luaT_netbox_transport_stop(struct lua_State *L)
 		netbox_transport_on_state_change(transport, L);
 	}
 	/* Cancel the worker fiber. */
-	if (transport->worker != NULL)
+	if (transport->worker != NULL) {
 		fiber_cancel(transport->worker);
+		/* Check if we cancelled ourselves. */
+		luaL_testcancel(L);
+	}
 	return 0;
 }
```
</details>

Gerold103 commented:
We will still need error copy in `luaT_netbox_request_discard()`. Otherwise looks good, vshard tests all pass!

locker commented:
> We will still need error copy in `luaT_netbox_request_discard()`. Otherwise looks good, vshard tests all pass!

Clearly you meant `netbox_request_push_result`, which returns `netbox_request::error` to the user:

https://github.com/tarantool/tarantool/blob/4c6d7b32e5449d2ea40c6b2b1ae859a07e493a59/src/box/lua/net_box.c#L308-L311

Yep, there we do need to copy an error, but this is issue exists since I rewritten net.box request in C so I guess we can leave it as is for now. I mentioned the fact in https://github.com/tarantool/tarantool/issues/6632#issuecomment-977602609 for the record.
